library/src/main/java/com/q42/q42stats/library/Q42Stats.kt

```diff
@@ -1,61 +1,58 @@
 package com.q42.q42stats.library
 
 import android.content.Context
+import androidx.annotation.AnyThread
+import androidx.annotation.WorkerThread
 import com.q42.q42stats.library.collector.AccessibilityCollector
 import com.q42.q42stats.library.collector.PreferencesCollector
 import com.q42.q42stats.library.collector.SystemCollector
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.GlobalScope
-import kotlinx.coroutines.launch
+import kotlinx.coroutines.*
 import java.io.Serializable
-import java.util.concurrent.atomic.AtomicBoolean
 
 const val TAG = "Q42Stats"
 
-private val MUTEX = Unit
-
 class Q42Stats(private val config: Q42StatsConfig) {
-    private val isRunning = AtomicBoolean(false)
 
     /* Collects stats and sends it to the server. This method is safe to be called from anywhere
-    in your app and will do nothing if it has already run before */
-    fun runAsync(context: Context) {
+    in your app and will do nothing if it is running or has already run before */
+    @AnyThread
+    fun runAsync(context: Context, coroutineScope: CoroutineScope = MainScope()) {
```

Comment:
I would instead provide two methods to the caller I think, so the caller can choose it's preference. One that launches on a MainScope(), without putting it as parameter in the method. The other a suspend method, that just switches context (withContext(Dispatchers.Main) ipv launch(Dispatchers.Main), so that one can use the suspend method and wait for it, if one likes. The scope as function param you can safely remove then.

Reply:
Most libs I could think of are written in Java, so don't provide information.

Did discuss with Daniello, but we could not think of a reason for a suspend fun since we could not think of:
- a reason to wait for a result because there isn't any
- a reason to cancel this job. (also; you could provide a custom Scope that you can cancel as a caller with the current api)

Passing the scope as a constructor param might be useful when the construction and usage happen in different parts of your codebase, but I also think that's unlikely. At the moment, I don't see a reason to change this implementation.

What is your reasoning?

Reply:
It's been a month: I don't remember. I did see it as tip in this video: https://360andev.com/2021-session-recordings/?vimeography_gallery=4&vimeography_video=579577623 at minute 4, but it's fine, this will work.